pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+	skymsg "github.com/skynetservices/skydns/msg"
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/endpoints"
+	kcache "k8s.io/kubernetes/pkg/client/cache"
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	kframework "k8s.io/kubernetes/pkg/controller/framework"
+	kselector "k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/util/validation"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+const (
+	kubernetesSvcName = "kubernetes"
+
+	// A subdomain added to the user specified domain for all services.
+	serviceSubdomain = "svc"
+
+	// A subdomain added to the user specified dmoain for all pods.
+	podSubdomain = "pod"
+
+	// Resync period for the kube controller loop.
+	resyncPeriod = 30 * time.Minute
+)
+
+type KubeDNS struct {
+	kubeClient *kclient.Client
+	// DNS domain name.
+	domain string
+	// A cache that contains all the endpoints in the system.
+	endpointsStore kcache.Store
+	// A cache that contains all the services in the system.
+	servicesStore     kcache.Store
+	cache             *TreeCache
+	domainPath        []string
+	eController       *kframework.Controller
+	serviceController *kframework.Controller
+}
+
+func NewKubeDNS(client *kclient.Client, domain string) *KubeDNS {
+	kd := &KubeDNS{
+		kubeClient: client,
+		domain:     domain,
+		cache:      NewTreeCache(),
+		domainPath: reverseArray(strings.Split(strings.TrimRight(domain, "."), ".")),
+	}
+	kd.setEndpointsStore()
+	kd.setServicesStore()
+	return kd
+}
+
+func (kd *KubeDNS) Start() {
+	go kd.eController.Run(wait.NeverStop)
+	go kd.serviceController.Run(wait.NeverStop)
+	// Wait synchronously for the Kubernetes service and add a DNS record for it.
+	// TODO (abshah) UNCOMMENT AFTER TEST COMPLETE
+	//kd.waitForKubernetesService()
+}
+
+func (kd *KubeDNS) waitForKubernetesService() (svc *kapi.Service) {
+	name := fmt.Sprintf("%v/%v", kapi.NamespaceDefault, kubernetesSvcName)
+	glog.Infof("Waiting for service: %v", name)
+	var err error
+	servicePollInterval := 1 * time.Second
+	for {
+		svc, err = kd.kubeClient.Services(kapi.NamespaceDefault).Get(kubernetesSvcName)
+		if err != nil || svc == nil {
+			glog.Infof("Ignoring error while waiting for service %v: %v. Sleeping %v before retrying.", name, err, servicePollInterval)
+			time.Sleep(servicePollInterval)
+			continue
+		}
+		break
+	}
+	return
+}
+
+func (kd *KubeDNS) GetCacheAsJSON() string {
+	json, _ := kd.cache.Serialize("")
+	return json
+}
+
+func (kd *KubeDNS) setServicesStore() {
+	// Returns a cache.ListWatch that gets all changes to services.
+	serviceWatch := kcache.NewListWatchFromClient(kd.kubeClient, "services", kapi.NamespaceAll, kselector.Everything())
+	kd.servicesStore, kd.serviceController = kframework.NewInformer(
+		serviceWatch,
+		&kapi.Service{},
+		resyncPeriod,
+		kframework.ResourceEventHandlerFuncs{
+			AddFunc:    kd.newService,
+			DeleteFunc: kd.removeService,
+			UpdateFunc: kd.updateService,
+		},
+	)
+}
+
+func (kd *KubeDNS) setEndpointsStore() {
```

Comment:
Oh, you're talking about recursive.  Yeah, we haven't enabled recursive up
until now.

On Thu, May 5, 2016 at 4:25 PM, Clayton Coleman <ccoleman@redhat.com> wrote:

> We've been running with 30s TTL in this setup for a long time.
>
> On Thu, May 5, 2016 at 4:24 PM, Clayton Coleman <ccoleman@redhat.com>
> wrote:
>
>> Why would we set skydns TTL to 10 mins anyway?
>>
>> On Thu, May 5, 2016 at 4:22 PM, Abhi Shah <notifications@github.com>
>> wrote:
>>
>>> In pkg/dns/dns.go
>>> <https://github.com/kubernetes/kubernetes/pull/23930#discussion_r62248366>
>>> :
>>>
>>> > +func (kd *KubeDNS) setServicesStore() {
>>> > +	// Returns a cache.ListWatch that gets all changes to services.
>>> > +	serviceWatch := kcache.NewListWatchFromClient(kd.kubeClient, "services", kapi.NamespaceAll, kselector.Everything())
>>> > +	kd.servicesStore, kd.serviceController = kframework.NewInformer(
>>> > +		serviceWatch,
>>> > +		&kapi.Service{},
>>> > +		resyncPeriod,
>>> > +		kframework.ResourceEventHandlerFuncs{
>>> > +			AddFunc:    kd.newService,
>>> > +			DeleteFunc: kd.removeService,
>>> > +			UpdateFunc: kd.updateService,
>>> > +		},
>>> > +	)
>>> > +}
>>> > +
>>> > +func (kd *KubeDNS) setEndpointsStore() {
>>>
>>> skydns caching is poor. It does not respect the TTL of the dns response.
>>>
>>> For example: if skydns TTL is set to 10 mins, and the upstream DNS
>>> response has a TTL of 30 seconds, skydns will still use the same response
>>> after 30 secs (upto 10 mins) without querying upstream.
>>>
>>> —
>>> You are receiving this because you were mentioned.
>>> Reply to this email directly or view it on GitHub
>>> <https://github.com/kubernetes/kubernetes/pull/23930/files/a33b1f831b93fa1402f82f687e83d0fb488d16cb..2b69fa8d68b059336a55ba56e9465e2a0c9e9fc9#r62248366>
>>>
>>
>>
>